lib/IDE/CodeCompletion.cpp

```diff
@@ -4124,6 +4126,79 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
     }
   }
 
+  void collectPrecedenceGroups() {
+    assert(CurrDeclContext);
+
+    auto addPrecedenceGroup = [&](PrecedenceGroupDecl *PGD) {
+      auto semanticContext =
+        getSemanticContext(PGD, DeclVisibilityKind::VisibleAtTopLevel);
+      CodeCompletionResultBuilder builder(
+        Sink, CodeCompletionResult::ResultKind::Declaration,
+        semanticContext, {});
+      builder.addTextChunk(PGD->getName().str());
+      builder.setAssociatedDecl(PGD);
+    };
+
+    auto collectPrecedenceGroups = [&](FileUnit *FU) {
+      switch (FU->getKind()) {
+      case FileUnitKind::SerializedAST: {
+        SmallVector<Decl *, 16> topLevelDecls;
+        cast<LoadedFile>(FU)->getTopLevelDecls(topLevelDecls);
```

Comment:
Sure. Roughly, where do we cache those?

Reply:
The cache is in `CodeCompletionCache` and you request results from it using `RequestedCachedModule`.  Since you'll need a way to only include precedence groups and not all symbols from the module, look at how we use the flag `OnlyType` inside that request; you could add an `OnlyPredenceGroups` that works similarly.

The requests are put in `Lookup.RequestedCachedResults` and then handled in `handleResultsAndModules` and `copyCodeCompletionResults`.